source/extensions/filters/http/cache/cache_insert_queue.h

```diff
@@ -13,37 +13,59 @@ namespace Cache {
 using OverHighWatermarkCallback = std::function<void()>;
 using UnderLowWatermarkCallback = std::function<void()>;
 using AbortInsertCallback = std::function<void()>;
-class CacheInsertChunk;
+class CacheInsertFragment;
 
+// This queue acts as an intermediary between CacheFilter and the cache
+// implementation extension. Having a queue allows CacheFilter to stream at its
+// normal rate, while allowing a cache implementation to run asynchronously and
+// potentially at a slower rate, without having to implement its own buffer.
+//
+// If the queue contains more than the "high watermark" for the buffer
+// (encoder_callbacks.encoderBufferLimit()), then a high watermark event is
+// sent to the encoder, which may cause the filter to slow down, to allow the
+// cache implementation time to catch up and avoid buffering significantly
+// more data in memory than the configuration intends to allow. When this happens,
+// the queue must drain to half the encoderBufferLimit before a low watermark
+// event is sent to resume normal flow.
+//
+// From the cache implementation's perspective, the queue ensures that the cache
+// receives data one piece at a time - no more data will be delivered until the
+// cache implementation calls the provided callback indicating that it is ready
+// to receive more data.
 class CacheInsertQueue {
 public:
   CacheInsertQueue(Http::StreamEncoderFilterCallbacks& encoder_callbacks,
                    InsertContextPtr insert_context, AbortInsertCallback abort);
   void insertHeaders(const Http::ResponseHeaderMap& response_headers,
                      const ResponseMetadata& metadata, bool end_stream);
-  void insertBody(const Buffer::Instance& chunk, bool end_stream);
+  void insertBody(const Buffer::Instance& fragment, bool end_stream);
   void insertTrailers(const Http::ResponseTrailerMap& trailers);
-  void takeOwnershipOfYourself(std::unique_ptr<CacheInsertQueue> self);
+  void setSelfOwned(std::unique_ptr<CacheInsertQueue> self);
```

Comment:
are we guaranteed that if we setSelfOwned the in flight operation will complete?  On the data plane side if we have an open connection it'll eventually time out and be closed down so I just want to make sure there's an equivalent contract for chunk fetching.

Reply:
mm, if it's going to be self owned I think we need to make it clear in caching APIs the cache is responsible for that.